File: ApiApplication/Entities/AuditLog.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace ApiApplication.Entities
{
    [Index(nameof(Timestamp), IsDescending = new[] { true })]
    [Index(nameof(TableName), nameof(Action))]
    [Index(nameof(TableName), nameof(EntityId))]
    [Index(nameof(UserId), nameof(Timestamp), IsDescending = new[] { false, true })]
    [Index(nameof(Action))]
    [Index(nameof(EntityId))]
    [Index(nameof(TableName))]
    [Index(nameof(UserId))]
    [Index(nameof(CreatedAt), IsDescending = new[] { true })]
    public class AuditLog
    {
        [Key]
        public Guid Id { get; set; }

        [Required]
        [StringLength(100)]
        public string TableName { get; set; } = string.Empty;

        [Required]
        [StringLength(20)]
        public string Action { get; set; } = string.Empty; // Create, Update, Delete

        [Required]
        [StringLength(50)]
        public string EntityId { get; set; } = string.Empty;

        [Column(TypeName = "text")]
        public string? OldValues { get; set; }

        [Column(TypeName = "text")]
        public string? NewValues { get; set; }

        [StringLength(1000)]
        public string? ChangedColumns { get; set; }

        [StringLength(50)]
        public string? UserId { get; set; }

        [StringLength(100)]
        public string? UserName { get; set; }

        [StringLength(45)]
        public string? IpAddress { get; set; }

        [StringLength(500)]
        public string? UserAgent { get; set; }

        [Required]
        public DateTime Timestamp { get; set; }

        [Required]
        public DateTime CreatedAt { get; set; }

        public DateTime? UpdatedAt { get; set; }

        public string? CreatedBy { get; set; }

        public string? UpdatedBy { get; set; }
    }
}

File: web-application/src/types/audit-log.ts
// Audit Log DTOs
export interface AuditLogDto {
  id: string;
  tableName: string;
  action: string; // "Create" | "Update" | "Delete"
  entityId: string;
  oldValues?: string | null;
  newValues?: string | null;
  changedColumns?: string | null;
  userId?: string | null;
  userName?: string | null;
  ipAddress?: string | null;
  userAgent?: string | null;
  timestamp: string;
  createdAt: string;
  updatedAt?: string | null;
  createdBy?: string | null;
  updatedBy?: string | null;
}

export interface AuditLogDetailDto {
  id: string;
  tableName: string;
  action: string;
  entityId: string;
  changedColumns?: string | null;
  oldValuesObject?: any;
  newValuesObject?: any;
  userId?: string | null;
  userName?: string | null;
  ipAddress?: string | null;
  userAgent?: string | null;
  timestamp: string;
}

export interface AuditLogQueryDto {
  tableName?: string | null;
  action?: string | null;
  userId?: string | null;
  entityId?: string | null;
  dateFrom?: string | null;
  dateTo?: string | null;
  searchKeyword?: string | null;
  page?: number;
  pageSize?: number;
}

export interface PagedResponse<T> {
  items: T[];
  page: number;
  pageSize: number;
  totalItems: number;
  totalPages: number;
}

export type AuditLogResponseDto = PagedResponse<AuditLogDto>;

export const AuditAction = {
  Create: "Create",
  Update: "Update",
  Delete: "Delete",
} as const;

export type AuditActionType = typeof AuditAction[keyof typeof AuditAction];
